### PR TITLE
Keep alive apps and scheduled-stop

### PR DESCRIPTION
**Pull request by coot**

### CODE REVIEW

File: ouroboros-network/src/Ouroboros/Network/KeepAlive.hs [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        Stop -> pure (SendMsgDone (pure ()))
+
+
+keepAliveServer
+  :: forall m.  Applicative m
+  => KeepAliveServer m ()
+keepAliveServer = KeepAliveServer {
+    recvMsgKeepAlive = pure keepAliveServer,
+    recvMsgDone      = pure ()
+  }
```
Comment (karknu) [inline ouroboros-network/src/Ouroboros/Network/KeepAlive.hs:59]: The exchange should be:
Client sends `MsgKeepAlive`
Server receives  `MsgKeepAlive`
Server sends `MsgKeepAliveResponse`
Client receives `MsgKeepAliveResponse`.

Comment (coot) [reply ouroboros-network/src/Ouroboros/Network/KeepAlive.hs:59]: That's exactly how the conversation will look like.

**Comment by coot** [comment]
@karknu this is req-resp model.  Every `KeepAliveInterval` the client sends `MsgKeepAlive` and the server replies with `MsgKeepAliveResponse` as soon as it receives client message.

File: ouroboros-network/src/Ouroboros/Network/Protocol/KeepAlive/Codec.hs [hunk 2]
```diff
@@ -53,3 +60,24 @@ codecKeepAlive = mkCodecCborLazyBS encodeMsg decodeMsg
...
+    sizeLimitForState (ClientAgency TokClient) = 1
+    sizeLimitForState (ServerAgency TokServer) = 1
+
+
+timeLimitsKeepAlive :: ProtocolTimeLimits KeepAlive
+timeLimitsKeepAlive = ProtocolTimeLimits { timeLimitForState }
+  where
+    timeLimitForState :: PeerHasAgency (pr :: PeerRole) (st :: KeepAlive)
+                      -> Maybe DiffTime
+    timeLimitForState (ClientAgency TokClient) = longWait
```
[the remark this answers is not in the document]
Comment (coot) [reply ouroboros-network/src/Ouroboros/Network/Protocol/KeepAlive/Codec.hs:82]: You're right, we don't need to be that restrictive here.

[the remark this answers is not in the document]
Comment (coot) [reply ouroboros-network/src/Ouroboros/Network/Protocol/KeepAlive/Codec.hs]: Fixed

**Comment by karknu** [comment]
> @karknu this is req-resp model. Every `KeepAliveInterval` the client sends `MsgKeepAlive` and the server replies with `MsgKeepAliveResponse` as soon as it receives client message.

You're right.

**Comment by coot** [comment]
bors merge

**Pull request merged**